source/_posts/2016/2016-12-24-how-to-avoid-inject-thanks-to-decorator-feature-in-nette.md

```diff
@@ -0,0 +1,254 @@
+
+The best use case is `AbstractBasePresenter`.  
+Let's say I need `Translator` service in all of my presenters.
+
+```language-php
+// app\Presenter\AbstractBasePresenter.php
+
+namespace App\Presenter;
+
+abstract class AbstractBasePresenter extends Nette\Application\UI\Presenter
+{
+    /**
+     * @inject
+     * @var Translator
+     */
+    public $translator;
+}
+```
+
+And I can use it `ProductPresenter` alongside with constructor injection
+ 
+```language-php
+// app/Presenter/ProductPresenter.php
+
+namespace App\Presenter;
+
+final class ProductPresenter extends AbstractBasePresenter
+{
+    /**
+     * @var ProductRepository
+     */
+    private $productRepository;
+    
+    public function __construct(ProductRepository $productRepository) 
+    {
+        $this->productRepository = $productRepository; 
+    }
+}
+```
+
+This is quite clean and easy to use, because presenters in Nette are [injected to by default](https://github.com/nette/application/blob/3165d3a8dab876f4364cdcba450a33ab0182049a/src/Bridges/ApplicationDI/ApplicationExtension.php#L111-L116).
+ 
+## Level up
+
+But what if we have other objects that:
+ 
+ - **inherit from abstract parent**
+ - needs **1 service available everywhere**
+
+2 common case pop to my mind:
+
+- `AbstractBaseRepository` for all our repositories
+- `AbstractBaseControl` for all our components
+
+Let's take the first one:
+
+```language-php
+// app/Repository/AbstractBaseRepository.php
+
+namespace App\Repository;
+
+use Doctrine\ORM\EntityManagerInterface;
+
+abstract class AbstractBaseRepository
+{
+    /**
+     * @var EntityManagerInterface
+     */
+    protected $entityManager;
+    
+    public function setEntityManager(EntityManagerInterface $entityManager)
+    {
+        $this->entityManager = $entityManager;
+    }
+}
+```
+
+And specific repository with some dependency:
+
+```language-php
+// app/Repository/ProductRepository.php
+
+namespace App\Repository;
+
+use App\Model\Product\ProductSorer;
+
+final ProductRepository extends AbstractBaseRepository
+{
+    /**
+     * @var ProductSorer
+     */
+    private $productSorter;
+    
+    public function __construct(ProductSorter $productSorter)
+    {
+        $this->productSorter = $productSorter;
+    }
+}
+```
+
+So our config would look like:
+
+```language-yaml
+# app/config/config.neon
+
+services:
+    - 
+        class: App\Repository\ProductRepository
+        setup:
+            - setEntityManager
+    # and for other repositories
+    - 
+        class: App\Repository\UserRepository
+        setup:
+            - setEntityManager
+    - 
+        class: App\Repository\CategoryRepository
+        setup:
+            - setEntityManager
+```           
+
+### SO much writing! 
+
+It is clean but with so much writing? Thanks, but no, thanks. Let's go back to `@inject`...
```

Comment:
No it's not exactly clean. Setter is a bit better than public property but still not exactly clean because the object still might not be fully initialized.

Also in my opinion inject* methods are bit better than set* methods because setter suggests that the service is not stateless (not the case here) where as inject* by convention are methods that should be called only once and immediately after constructor.

(Of course with inject* methods you don't need the decorator to add "setup" but to add the inject tag.)